packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -959,3 +921,1328 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+  // key reparenting in the same frame but N is usually a small number.
+  void _add(_OverlayEntryLocation child) {
+    assert(_debugNotDisposed());
+    assert(!_sortedChildren.contains(child));
+    _OverlayEntryLocation? insertPosition = _sortedChildren.isEmpty ? null : _sortedChildren.last;
+    while (insertPosition != null && insertPosition._showTimestamp > child._showTimestamp) {
+      insertPosition = insertPosition.previous;
+    }
+    if (insertPosition == null) {
+      _sortedChildren.addFirst(child);
+    } else {
+      insertPosition.insertAfter(child);
+    }
+    assert(_sortedChildren.contains(child));
+
+    if (child._overlayChildRenderBox != null) {
+      _theatre.markNeedsPaint();
+      _theatre.markNeedsCompositingBitsUpdate();
+      _theatre.markNeedsSemanticsUpdate();
+    }
+  }
+
+  void _remove(_OverlayEntryLocation child) {
+    final bool wasInCollection = _sortedChildren.remove(child);
+    assert(wasInCollection);
+    if (child._overlayChildRenderBox != null) {
+      _theatre.markNeedsPaint();
+      _theatre.markNeedsCompositingBitsUpdate();
+      _theatre.markNeedsSemanticsUpdate();
+    }
+  }
+
+  _OverlayEntryLocation? _childAfter(_OverlayEntryLocation? child) {
+    if (_sortedChildren.isEmpty) {
+      return null;
+    }
+    final _OverlayEntryLocation? candidate = child == null ? _sortedChildren.first : child.next;
+    return candidate != null ? candidate._overlayChildRenderBox != null ? candidate : _childAfter(candidate) : null;
+  }
+
+  _OverlayEntryLocation? _childBefore(_OverlayEntryLocation? child) {
+    if (_sortedChildren.isEmpty) {
+      return null;
+    }
+    final _OverlayEntryLocation? candidate = child == null ? _sortedChildren.last : child.previous;
+    return candidate != null ? candidate._overlayChildRenderBox != null ? candidate : _childBefore(candidate) : null;
+  }
+
+  static _ChildIterator? _mapIterator(_Iterator<_OverlayEntryLocation>? fromIterator) {
+    if (fromIterator == null) {
+      return null;
+    }
+
+    _RenderDeferredLayoutBox? iter() {
+      final _OverlayEntryLocation? location = fromIterator();
+      if (location == null) {
+        return null;
+      }
+      return location._overlayChildRenderBox ?? iter();
+    }
+
+    return iter;
+  }
+
+  _Iterator<_OverlayEntryLocation>? _forwardIterator() {
+    _OverlayEntryLocation? child = _childAfter(null);
+    if (child == null) {
+      return null;
+    }
+    bool isInitial = true;
+    return () {
+      if (isInitial) {
+        assert(child != null);
+        isInitial = false;
+        return child;
+      }
+      return child == null ? null : child = _childAfter(child);
+    };
+  }
+
+  _Iterator<_OverlayEntryLocation>? _backwardIterator() {
+    _OverlayEntryLocation? child = _childBefore(null);
+    if (child == null) {
+      return null;
+    }
+    bool isInitial = true;
+    return () {
+      if (isInitial) {
```

Comment:
What happen if the children change while caller using this class?

Reply:
ah it is used during paint and hit test, let's add an assert or comment to explain that

Reply:
Yeah the comments are in a different place: https://github.com/flutter/flutter/pull/105335/files#diff-6ea34306ad31c88d41a39d24aaad65a2ef227ebbee8919697b3411ff2ef0f897R1609-R1614

I'll add them here too.

Reply:
added.